File: cvicenia/bonus01/bonus01Test.cpp
#include <iostream>
#include <string>
#include "Formula.h"

struct Case {
	Case(const Valuation &v_, bool result_)
		: v(v_)
		, result(result_)
	{}
	Valuation v;
	bool result;
};

/**
 * Pekny vypis ohodnotenia
 */
std::ostream& operator<< (std::ostream& stream, const Valuation &v)
{
	stream << "{ ";
	for(auto p : v) {
		stream << p.first << ": " << p.second << " ";
	}
	stream << "}";
	return stream;
}

class Tester {
	int m_tested = 0;
	int m_passed = 0;
public:
	template<typename T>
	void compare(const T &result, const T &expected, const std::string &msg)
	{
		m_tested++;
		if (result == expected) {
			m_passed++;
		}
		else {
			std::cerr << "    Failed: " << msg << ":" << std::endl;
			std::cerr << "      got " << result << " expected: " << expected << std::endl;
		}
	}

	// typeid(x).name() is not that reliable
	std::string formulaType(Formula *a)
	{
		if (dynamic_cast<Variable*>(a))
			return "Variable";
		if (dynamic_cast<Negation*>(a))
			return "Negation";
		if (dynamic_cast<Disjunction*>(a))
			return "Disjunction";
		if (dynamic_cast<Conjunction*>(a))
			return "Conjunction";
		if (dynamic_cast<Implication*>(a))
			return "Implication";
		if (dynamic_cast<Equivalence*>(a))
			return "Equivalence";
		if (dynamic_cast<Formula*>(a))
			return "Formula";
		return "UNKNOWN";
	}

	bool compareFormulas(Formula *a, Formula *b)
	{
		if (formulaType(a) != formulaType(b)) {
			std::cerr << "    Failed: Formulas do not match" << std::endl;
			std::cerr << "      parsed: " << formulaType(a) << " expected: " << formulaType(b) << std::endl;
			return false;
		}
		if (a->subf().size() != b->subf().size()) {
			std::cerr << "    Failed: different subformula list sizes" << std::endl;
			std::cerr << "      parsed: " << a->subf().size() << " expected: " << b->subf().size() << std::endl;
			return false;
		}
		for(int i=0; i<a->subf().size(); ++i) {
			if (!compareFormulas(a->subf()[i], b->subf()[i]))
				return false;
		}
		return true;
	}

	void test(Formula *f, std::string str, const std::vector<Case> &cases)
	{
		std::cerr << "Testing " << str << std::endl;

		Formula *parsed = Formula::parse(str);
		compare(parsed->toString(), str, "Formula::parse(XXX).toString() == XXX");

		// This expects subf() to work correctly, if it doesn't
		// we still will catch the difference in toplevel formulas
		// or isSatisfieds below.
		compareFormulas(parsed, f);

		for (const auto &c : cases) {
			std::cerr << "Valuation " << c.v << std::endl;
			compare(parsed->isSatisfied(c.v), c.result, "isSatisfied");
		}

		delete parsed;
		delete f;
	}
	void status()
	{
		std::cerr << "TESTED " << m_tested << std::endl;
		std::cerr << "PASSED " << m_passed << std::endl;
		std::cerr << ( m_tested == m_passed ? "OK" : "ERROR" ) << std::endl;
	}
};


int main()
{
	Tester t;

	Valuation va1, va2;
	va1["a"] = true;  va2["a"] = false;
	t.test(
			new Variable("a"),
			"a",
			{
				Case(va1, true),
				Case(va2, false),
			});

	t.test(
			new Negation(new Variable("a")), "-a",
			{
				Case(va1, false),
				Case(va2, true),
			});


	std::vector<Valuation> valuations2;
	{ Valuation v; v["a"] = false; v["b"] = false ; valuations2.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = true  ; valuations2.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = false ; valuations2.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = true  ; valuations2.push_back(v); }

	t.test(
			new Conjunction( { new Variable("a"), new Variable("b") } ),
			"(a&b)",
			{
				Case(valuations2[0], false),
				Case(valuations2[1], false),
				Case(valuations2[2], false),
				Case(valuations2[3], true),
			});

	t.test(
			new Disjunction( { new Variable("a"), new Variable("b") } ),
			"(a|b)",
			{
				Case(valuations2[0], false),
				Case(valuations2[1], true),
				Case(valuations2[2], true),
				Case(valuations2[3], true),
			});

	t.test(
			new Implication( new Variable("a"), new Variable("b") ),
			"(a->b)",
			{
				Case(valuations2[0], true),
				Case(valuations2[1], true),
				Case(valuations2[2], false),
				Case(valuations2[3], true),
			});

	t.test(
			new Equivalence( new Variable("a"), new Variable("b") ),
			"(a<->b)",
			{
				Case(valuations2[0], true),
				Case(valuations2[1], false),
				Case(valuations2[2], false),
				Case(valuations2[3], true),
			});

	t.test(
			new Disjunction({
				new Negation(new Implication(new Variable("a"),new Variable("b"))),
				new Negation(new Implication(new Variable("b"),new Variable("a")))
			}),
			"(-(a->b)|-(b->a))",
			{
				Case(valuations2[0], false),
				Case(valuations2[1], true),
				Case(valuations2[2], true),
				Case(valuations2[3], false),
			});

	std::vector<Valuation> valuations3;
	{ Valuation v; v["a"] = false; v["b"] = false, v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = false, v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = true , v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = true , v["c"] = false ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = false, v["c"] = true  ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = false, v["c"] = true  ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = false; v["b"] = true , v["c"] = true  ; valuations3.push_back(v); }
	{ Valuation v; v["a"] = true ; v["b"] = true , v["c"] = true  ; valuations3.push_back(v); }

	t.test(
			new Conjunction({
				new Implication(new Variable("a"),new Variable("b")),
				new Implication(new Negation(new Variable("a")),new Variable("c"))
			}),
			"((a->b)&(-a->c))",
			{
				Case(valuations3[0], false),
				Case(valuations3[1], false),
				Case(valuations3[2], false),
				Case(valuations3[3], true),
				Case(valuations3[4], true),
				Case(valuations3[5], false),
				Case(valuations3[6], true),
				Case(valuations3[7], true),
			});

	t.test(
			new Equivalence(
				new Conjunction({
					new Variable("a"),
					new Negation(new Variable("b"))
				}),
				new Disjunction({
					new Variable("a"),
					new Implication(
						new Variable("b"),
						new Variable("a")
					)
				})
			),
			"((a&-b)<->(a|(b->a)))",
			{
				Case(valuations2[0], false),
				Case(valuations2[1], true),
				Case(valuations2[2], true),
				Case(valuations2[3], false),
			});

	t.status();
	return 0;
}

File: cvicenia/bonus01/bonus01Test.py
#!/bin/env python

"""
  Testovaci program pre kniznicu formula.py
"""

from formula import Formula, Variable, Negation, Conjunction, Disjunction, Implication, Equivalence

class Tester(object):
    def __init__(self):
        self.tested = 0
        self.passed = 0
    def compare(self, result, expected, msg):
        self.tested += 1
        if result == expected:
            self.passed += 1
        else:
            print("    Failed: %s:" %  msg)
            print("      got %s  expected %s" % (repr(result), repr(expected)))

    def compareFormulas(self, f1, f2):
        if type(f1) != type(f2):
            print("    Failed: Formulas do not match")
            print("      parsed: %s   expected: %s" % (type(f1), type(f2)))
            return False
        if len(f1.subf()) != len(f2.subf()):
            print("    Failed: Different subformula lists")
            print("      parsed: %s   expected: %s" % (f1.subf(), f2.subf()))
            return False
        for a,b in zip(f1.subf(), f2.subf()):
            if not self.compareFormulas(a,b):
                return False
        return True

    def test(self, formula, string, cases):
        print("Testing %s" % (string))

        parsed = Formula.parse(string)

        self.compare(parsed.toString(), string, "Formula.parse(XXX).toString() == XXX")

        # This expects subf() to work correctly, if it doesn't
        # we still will catch the difference in toplevel formulas
        # or isSatisfieds below.
        self.compareFormulas(parsed, formula)

        for valuation, result in cases:
            print(" valuation %s:" % (repr(valuation),))
            self.compare(parsed.isSatisfied(valuation), result, "isSatisfied")


    def status(self):
        print("TESTED %d" % (self.tested,))
        print("PASSED %d" % (self.passed,))
        if self.tested == self.passed:
            print("OK")
        else:
            print("ERROR")

t = Tester()

t.test(
        Variable('a'), 'a',
        [
            ({'a':True}, True),
            ({'a':False}, False),
        ])

t.test(
        Negation(Variable('a')), '-a',
        [
            ({'a':True}, False),
            ({'a':False}, True),
        ])


valuations2 = [{ 'a': False, 'b': False },
               { 'a': False, 'b': True  },
               { 'a': True , 'b': False },
               { 'a': True , 'b': True  }]

t.test(
        Conjunction( [ Variable('a'), Variable('b') ] ),
        '(a&b)',
        [
            (valuations2[0], False),
            (valuations2[1], False),
            (valuations2[2], False),
            (valuations2[3], True),
        ])

t.test(
        Disjunction( [ Variable('a'), Variable('b') ] ),
        '(a|b)',
        [
            (valuations2[0], False),
            (valuations2[1], True),
            (valuations2[2], True),
            (valuations2[3], True),
        ])

t.test(
        Implication( Variable('a'), Variable('b') ),
        '(a->b)',
        [
            (valuations2[0], True),
            (valuations2[1], True),
            (valuations2[2], False),
            (valuations2[3], True),
        ])

t.test(
        Equivalence( Variable('a'), Variable('b') ),
        '(a<->b)',
        [
            (valuations2[0], True),
            (valuations2[1], False),
            (valuations2[2], False),
            (valuations2[3], True),
        ])

t.test(
        Disjunction([
            Negation(Implication(Variable('a'),Variable('b'))),
            Negation(Implication(Variable('b'),Variable('a')))
        ]),
        '(-(a->b)|-(b->a))',
        [
            (valuations2[0], False),
            (valuations2[1], True),
            (valuations2[2], True),
            (valuations2[3], False),
        ])

valuations3 = [{ 'a': False, 'b': False, 'c': False },
               { 'a': True , 'b': False, 'c': False },
               { 'a': False, 'b': True , 'c': False },
               { 'a': True , 'b': True , 'c': False },
               { 'a': False, 'b': False, 'c': True  },
               { 'a': True , 'b': False, 'c': True  },
               { 'a': False, 'b': True , 'c': True  },
               { 'a': True , 'b': True , 'c': True  }]

t.test(
        Conjunction([
            Implication(Variable('a'),Variable('b')),
            Implication(Negation(Variable('a')),Variable('c'))
        ]),
        '((a->b)&(-a->c))',
        [
            (valuations3[0], False),
            (valuations3[1], False),
            (valuations3[2], False),
            (valuations3[3], True),
            (valuations3[4], True),
            (valuations3[5], False),
            (valuations3[6], True),
            (valuations3[7], True),
        ])

t.test(
        Equivalence(
            Conjunction([
                Variable('a'),
                Negation(Variable('b'))
            ]),
            Disjunction([
                Variable('a'),
                Implication(
                    Variable('b'),
                    Variable('a')
                )
            ])
        ),
        '((a&-b)<->(a|(b->a)))',
        [
            (valuations2[0], False),
            (valuations2[1], True),
            (valuations2[2], True),
            (valuations2[3], False),
        ])

t.status()

# vim: set sw=4 ts=4 sts=4 et :
